### PR TITLE
`github:pr` command

### PR DESCRIPTION
**Pull request by cmuench**
Magerun pull-request check-list:

- [X] Pull request against develop branch (if not, just close and create a new one against it)
- [ ] README.md reflects changes (if any)
- [ ] phar fuctional test (in tests/phar-test.sh)

Summary: Add a new command to work with Github PRs

Features of the new command

- Dump info of the PR
- Support Mage-OS
- Create a patch file from PR number which can be applied to Magento

### CODE REVIEW

**Comment by cmuench** [comment]
> Very cool.
> 
> Does it support auto discovery i.e. adding `@package` and `@level` tags in the patch file?

I currently replace only paths.
Can you describe what you need? I am not aware of the feature.

**Comment by amenk** [comment]
See https://github.com/vaimo/composer-patches#usage-embedded-metadata

Basically the package to patch and the level of directories to remove is added to the patch file. Together with the search path defined once in the composer.json, the patches should just work without any additional manual changes, if this is implemented

File: src/N98/Magento/Command/Github/PullRequestCommand.php [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+                InputArgument::REQUIRED,
+                'Pull Request Number'
+            )
+            ->addOption('repository', 'r', InputOption::VALUE_OPTIONAL, 'Repository to fetch from', 'magento/magento2')
+            ->addOption('mage-os', null, InputOption::VALUE_NONE, 'Shortcut option to use the mage-os/mageos-magento2 repository.')
+            ->addOption('patch', 'd', InputOption::VALUE_NONE, 'Download patch and prepare it for applying')
+            ->addOption('diff', null, InputOption::VALUE_NONE, 'Raw diff download')
+            ->addOption('json', null, InputOption::VALUE_NONE, 'Show pull request data as json')
+            ->setDescription('Download patch from github merge request <comment>(experimental)</comment>');
+    }
```
Comment (hostep) [inline src/N98/Magento/Command/Github/PullRequestCommand.php:34]: Would be nice to have an option to specify to which directory it should write the patch file(s)

File: src/N98/Magento/Command/Github/PullRequestCommand.php [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+            str_replace('/', '-', $prData['base']['repo']['full_name'])
+        );
+
+        chdir(OperatingSystem::getCwd());
+        if (file_put_contents($filename, $patchFileContent) === false) {
+            throw new \RuntimeException('Could not write patch file');
+        }
+
+        $output->writeln(sprintf('<info>Patch file created:</info> <comment>%s</comment>', $filename));
+    }
```
Comment (hostep) [inline src/N98/Magento/Command/Github/PullRequestCommand.php:123]: Does this method handle generating multiple patch files, for each module separately? It's a bit hard to check by just reading the code, but I think it doesn't?

For example: https://github.com/magento/magento2/pull/34360
Should (in my opinion, but some people disagree) be splitted in 2 patches, for:
- `magento/magento2-base`
- `magento/module-quote-graph-ql`

Comment (cmuench) [reply src/N98/Magento/Command/Github/PullRequestCommand.php:123]: Right. PR is not ready to merge and some stuff is currently not handled. I guess we have also to manage the language packs here.

The magento/framework-message-queue package was initially part of the Enterprise Edition. Maybe that's why it's not part of the framework.

Is splitting in different patch files a good idea?

Comment (hostep) [reply src/N98/Magento/Command/Github/PullRequestCommand.php:123]: Splitting is in my eyes necessary because you need to specify to which module a patch should apply, and if you specify from one module that some patch should apply to another module it will not work if you change the paths. 

Unless you don't change the paths and then specify a very high patch level (I think [it's 4](https://github.com/cweagans/composer-patches/pull/136) for Magento?).
Also found this super old discussion from years ago about this: https://github.com/cweagans/composer-patches/issues/76#issuecomment-394078040

I'd say, give it a try with a complex PR like I specified above and see how well it works.

Comment (cmuench) [reply src/N98/Magento/Command/Github/PullRequestCommand.php:123]: If someone uses ": "cweagans/composer-patches" then... we e.g. do not use it in Magento projects. The Quality Patches are also not splitted.

Comment (cmuench) [reply src/N98/Magento/Command/Github/PullRequestCommand.php:123]: My current plan is not to split. Splitting sounds like a lot of possible issues.

Comment (cmuench) [reply src/N98/Magento/Command/Github/PullRequestCommand.php:123]: At netz98 we have an own patch repository with a apply-patch.sh script (in the patch repo). This is then called via post-install-script/post-update-script in the `composer.json`.
Super simple solution.
Similar to the Quality Patches of Adobe.

For n98-magerun2 we should find a useful way for everyone. We use `cweagans/composer-patches` in other PHP projects, too. For Magento projects I do not like this splitting. IMHO it's a over-complication.
This does not mean that we should not support it (if it's possible).

I have not used the `vaimo/composer-patches` package. On the first look it looks for me that a standard (not splitted patch) can be used.

**Comment by cmuench** [comment]
@hostep The Message Queue edge case is now handled and we have support for `app/i18n`.
 I refactored the whole stuff by creating processors which can easily be tested (Tests are already provided).
